server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -3033,22 +2991,33 @@ private static boolean isQueued(@Nullable ShardSnapshotStatus status) {
         private final ClusterState currentState;
 
         // updates outstanding to be applied to existing snapshot entries
-        private final List<ShardSnapshotUpdate> unconsumedUpdates;
+        private final Map<String, List<ShardSnapshotUpdate>> unconsumedUpdates;
 
         // updates that were used to update an existing in-progress shard snapshot
         private final Set<ShardSnapshotUpdate> executedUpdates = new HashSet<>();
 
         SnapshotShardsUpdateContext(ClusterState currentState, List<ShardSnapshotUpdate> updates) {
             this.currentState = currentState;
-            unconsumedUpdates = new ArrayList<>(updates);
+            unconsumedUpdates = new HashMap<>();
+            for (ShardSnapshotUpdate update : updates) {
+                unconsumedUpdates.computeIfAbsent(update.snapshot.getRepository(), r -> new ArrayList<>()).add(update);
+            }
         }
 
         ClusterState computeUpdatedState() {
-            final List<SnapshotsInProgress.Entry> oldEntries = currentState.custom(SnapshotsInProgress.TYPE, SnapshotsInProgress.EMPTY)
-                .entries();
-            final List<SnapshotsInProgress.Entry> newEntries = new ArrayList<>(oldEntries.size());
-            for (SnapshotsInProgress.Entry entry : oldEntries) {
-                newEntries.add(applyToEntry(entry));
+            final SnapshotsInProgress existing = currentState.custom(SnapshotsInProgress.TYPE, SnapshotsInProgress.EMPTY);
+            SnapshotsInProgress updated = existing;
+            for (Map.Entry<String, List<ShardSnapshotUpdate>> updates : unconsumedUpdates.entrySet()) {
+                final String repoName = updates.getKey();
+                final List<SnapshotsInProgress.Entry> oldEntries = existing.forRepo(repoName);
+                if (oldEntries.isEmpty()) {
+                    continue;
+                }
+                final List<SnapshotsInProgress.Entry> newEntries = new ArrayList<>(oldEntries.size());
+                for (SnapshotsInProgress.Entry entry : oldEntries) {
+                    newEntries.add(applyToEntry(entry, updates.getValue()));
+                }
+                updated = updated.withUpdatedEntriesForRepo(repoName, newEntries);
```

Comment:
This makes a new map for every repo with active updates. Often that's not many but I've certainly heard folks talking about having lots of repositories. Could we instead create a map of repository contexts (between the top-level update context and the individual entry contexts) and use it to do a one-off map creation instead?

Reply:
I think I'd like to skip this one :) The planned follow-up here will redo the construction and updating of `SnapshotsInProgress` anyway to finally get away from all this magical building and updating of the entries list in just the right order so this is just going to be pretty short lived and good enough for that. Creating a few extra maps here, even if you have 10 repos should be completely irrelevant relative to the other work we're doing in this state update cost wise.